tests/lang/s02_core/s99_coroutines/t04_launch_coroutine.nim

```diff
@@ -0,0 +1,27 @@
+
+## To launch a coroutine, the built-in ``launch`` routine is used. As its
+## single parameter, it expects the invocation of a coroutine.
+##
+## What happens is, in the following order:
+## 1. a managed heap cell of the coroutine's internal environment type is
+##    allocated
+## 2. the provided arguments are captured in the internal environment, either
+##    by copying or moving, depending on the parameter
+## 3. the status of the couroutine instance is set to the "suspended" state
+##
+## The instantiated coroutine is returned as the built-in ``Coroutine[T]``
+## type (where `T` is the return type of the coroutine), which is a
+## polymorphic ``ref`` type. A coroutine's internal environment type is always
+## derived from the ``Coroutine[T]`` type.
+##
+## The body of the coroutine is not executed yet.
+
+proc coro(x: int) {.coroutine.} =
+  discard
+
+# every routine invocation syntax is valid for ``launch``
+let instance = launch coro(1)
```

Comment:
`launch` is indeed quite foreign, in that it essentially acts as a keyword that changes the meaning of its operand (i.e., the `coro(1)` isn't a routine invocation).

I like the idea of using `yield` to replace what is currently `resume`/`suspend`, but it would clash with iterators, and having whether `yield` returns something or not depend on the enclosing context (coroutine vs. anything else) could be a problem, I think.

If `coro(1)` is the coroutine constructor, how would `t19_automatic_trampoline` work? (I'm not entirely sure the `trampoline` expansion is a good idea, but it would would allow treating a coroutine like a normal procedure without any special syntax)

Reply:
> If `coro(1)` is the coroutine constructor, how would `t19_automatic_trampoline` work? (I'm not entirely sure the `trampoline` expansion is a good idea, but it would would allow treating a coroutine like a normal procedure without any special syntax)

I'm not sure about it, but my first thought is perhaps the compiler treats it within `discardCheck` to automatically execute coroutines. I'm not sure how satisfying that is, however.

Reply:
Python and C# have `yield from` and `yield return` respectively. They delegate control from the current coroutine to a sub-coroutine.

Reply:
`coro(1)` is now the construction expression itself, making `launch` obsolete. Thanks for the input, everyone.